### PR TITLE
Improve e-mail notifications

### PR DESCRIPTION
**Pull request by nnhimes**
Reopening of #2109 E2142. Improve e-mail notifications with some merge conflicts resolved.
To close issue #111 primarily

Need to also look into merging #2089 which would change some of this code. I think this can be merged first since it changes less functionality

### CODE REVIEW

**Comment by nnhimes** [comment]
[Wiki](https://expertiza.csc.ncsu.edu/index.php/CSC/ECE_517_Fall_2021_-_E2142._Improve_e-mail_notifications)
Tries to fix issues #717 #345 #111

File: app/views/mailer/partials/_updated_submission_since_review_html.html.erb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+Hi <%= @first_name %>,</br>
+<p>
+  A <%=@type%> you were reviewing for Assignment '<%= @obj_name %>' has been updated since you last review.<br>
+  Hence, to update your review please click on the following link.<br>
+  <% if @link != nil %><a href="<%= @link %>"><%= @link %></a><% end %>
```
Comment (johnbumgardner) [inline app/views/mailer/partials/_updated_submission_since_review_html.html.erb:5]:
```suggestion
  <% if @link %><a href="<%= @link %>"><%= @link %></a><% end %>
```

File: app/views/mailer/partials/_updated_submission_since_review_plain.html.erb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+Hi <%= @first_name %>,
+
+A <%=@type%> you were reviewing for Assignment '<%= @obj_name %>' has been updated since you last review.
+Hence, to update your review please click on the following link.
+<% if @link != nil %><%= @link %><% end %>
```
Comment (johnbumgardner) [inline app/views/mailer/partials/_updated_submission_since_review_plain.html.erb:5]:
```suggestion
<% if @link %><%= @link %><% end %>
```

File: spec/features/assignment_submission_spec.rb [hunk 3]
```diff
@@ -135,4 +135,29 @@ def signup_topic
     click_on 'Upload file'
     expect(page).to have_content 'File type error'
   end
+
+  describe "notification of reviewers" do
+
+    it "does not notify the reviewer/s when no reviews have been submitted" do
```
Comment (johnbumgardner) [inline spec/features/assignment_submission_spec.rb:141]: weird whitespace inconsistent with rest of file

File: spec/models/participant_spec.rb [hunk 4]
```diff
@@ -151,4 +152,18 @@
       expect(Participant.sort_by_name(unsorted)).to eq(sorted)
     end
   end
+
+  describe 'check if email is being sent or not' do
+
+    it 'participants assignment reviewers are sent email for a new submission' do
+
```
Comment (johnbumgardner) [inline spec/models/participant_spec.rb:159]: unnecessary whitespace

File: spec/workers/sidekiq_mail_worker_spec.rb [hunk 5]
```diff
@@ -7,21 +7,32 @@
   before(:each) do
     allow(Assignment).to receive(:find).with('1').and_return(assignment)
     allow(Participant).to receive(:where).with(parent_id: '1').and_return([participant])
+    allow(User).to receive(:where).with(email: "psingh22@ncsu.edu").and_return([user])
+    allow(Participant).to receive(:find).with(user_id: '1', parent_id: '1').and_return([participant])
```
Comment (johnbumgardner) [inline spec/workers/sidekiq_mail_worker_spec.rb:11]: i dont think that find works this way. find only accepts id. this should be where i think

File: spec/workers/sidekiq_mail_worker_spec.rb [hunk 6]
```diff
@@ -7,21 +7,32 @@
...
-      # expect(email.subject).to eq('Message regarding teammate review for assignment ' + assignment.name)
-      expect(email.subject).to eq('Your Expertiza account and password has been created')
+      expect(email.from[0]).to eq("expertiza.development@gmail.com")
+      expect(email.bcc[0]).to eq("psingh22@ncsu.edu")
+      expect(email.subject).to eq("Message regarding teammate review for assignment no assignment")
+      expect(email.body).to eq("This is a reminder to complete teammate review for assignment no assignment.\nPlease follow the link: http://expertiza.ncsu.edu/student_task/view?id=1\nDeadline is 2018-12-31 00:00:01. If you have already done the teammate review, then please ignore this mail.")
     end
 
-    it 'should expect the queue size of one' do
+    it "should expect the queue size of one" do
```
Comment (johnbumgardner) [inline spec/workers/sidekiq_mail_worker_spec.rb:35]: these should all be single quotes in this file

File: app/mailers/mail_worker.rb [hunk 7]
```diff
@@ -36,15 +36,26 @@ def perform(assignment_id, deadline_type, due_at)
...
+
+    # Defining the body of mail
+    body = "This is a reminder to complete #{deadline_type} for assignment #{assignment.name}.\n"
 
     emails.each do |mail|
+      # Since the email is not unique fetching the first instance in results returned when searching via email
+      user = User.where(email: mail).first
+
+      # Finding the mapping between participant and assignment so that it can be sent as a query param
+      participant_assignment_id = Participant.find(user_id: user.id.to_s, parent_id: self.assignment_id.to_s).first.id
```
Comment (johnbumgardner) [inline app/mailers/mail_worker.rb:48]: i believe this ought to be a where. find expects id

File: app/models/participant.rb [hunk 8]
```diff
@@ -73,6 +73,33 @@ def topic_name
...
+  # send email to team's reviewers in case a new submission is made
+  def mail_assigned_reviewers
+    # Find review mappings for the work done by this participant's team
+    mappings = ResponseMap.where(reviewed_object_id: self.assignment.id,
+                                 reviewee_id: self.team.id,
+                                 type: 'ReviewResponseMap')
+    unless mappings.nil?
+      mappings.each do |mapping|
+        reviewer = mapping.reviewer.user
+        Mailer.sync_message(
```
Comment (johnbumgardner) [inline app/models/participant.rb:85]: this mail creation code should be somewhere else. single responsibility principle.

**Review by johnbumgardner — CHANGES_REQUESTED** [review]
fix these issues and manually test

File: app/models/participant.rb (hunk 8 above)
Comment (nnhimes) [reply app/models/participant.rb:85]: Not sure where to move this mail creation/delivery code. Similar code is handled in models, controllers, and helpers right now. Maybe all of it could be consolidated into mailer.rb eventually?
image

Comment (nnhimes) [reply app/models/participant.rb:85]: Never mind on this actually, I'll cut down on the actual creation lines in participant by making a new method in mailer_helper.rb. Then participant will still send the mail

**Review by johnbumgardner — APPROVED** [review]
lgtm if manually tested

**Comment by nnhimes** [comment]
**Testing:**

- _Issue 1 (Import from CSV sends welcome email)_: 
Cannot get past import stage to see if email sends. When trying to import a participant to a course using this format: 

image
This error is given and the participant is not added:
image
The wiki page has a screenshot of the welcome email working, so this might be a bug introduced by recent localization or user creation password removal PRs. 

- _Issue 2 (Emails should be sent to reviewer when a new submission is made on an already-reviewed assignment)_: 
This issue has a pretty obscure set of steps to achieve the email send (see wiki page test steps at bottom). I'm not sure this combination of being able to submit in the review period and review during the submission period is ever used in the class. 
I'm going to test with the normal submit/review/submit/review structure instead.
image
It sends a review link that would only be helpful if the submission period allowed reviewed. 

- _Issue 3 (Adding relevant links to reminder emails)_ : 
Not receiving any reminder emails with deadlines set to 1 hr in the future.
